File: core_api/routes/stocks.py
"""Routes for stock data and analysis."""

from fastapi import APIRouter, HTTPException, Query
from typing import List, Dict, Any, Optional
import psycopg2
from datetime import datetime, timedelta
import os

router = APIRouter()


def get_db_connection():
    """Get PostgreSQL connection."""
    dsn = os.getenv(
        "TRADING_CHITTI_PG_DSN",
        "postgresql://hariprasath@localhost:6432/trading_chitti"
    )
    return psycopg2.connect(dsn)


@router.get("/top-gainers")
async def get_top_gainers(limit: int = Query(20, le=100)) -> List[Dict[str, Any]]:
    """Get top gainers based on real-time price changes from Zerodha Kite WebSocket.

    Returns stocks with highest positive price changes today.
    """
    try:
        conn = get_db_connection()
        cur = conn.cursor()

        # Get real-time gainers from WebSocket data
        # Calculate change_percent from close vs last_price
        cur.execute("""
            SELECT
                i.tradingsymbol as symbol,
                i.name,
                r.last_price::float as price,
                CASE
                    WHEN r.close > 0 THEN
                        ((r.last_price - r.close) / r.close * 100)::float
                    ELSE 0
                END as change,
                CASE
                    WHEN r.close > 0 THEN
                        LEAST(ABS((r.last_price - r.close) / r.close * 100) / 10.0, 1.0)
                    ELSE 0.5
                END as confidence
            FROM md.realtime_prices r
            JOIN md.instrument_tokens i ON i.instrument_token = r.instrument_token
            WHERE i.exchange = 'NSE'
              AND i.instrument_type = 'EQ'
              AND r.updated_at > NOW() - INTERVAL '10 minutes'
              AND r.close > 0
              AND r.last_price > r.close
            ORDER BY change DESC
            LIMIT %s
        """, (limit,))

        results = cur.fetchall()
        cur.close()
        conn.close()

        if not results:
            return []

        return [
            {
                "symbol": row[0],
                "name": row[1] or row[0],
                "change": round(float(row[3]) if row[3] else 0.0, 2),
                "confidence": round(float(row[4]) if row[4] else 0.0, 2),
                "price": round(float(row[2]) if row[2] else 0.0, 2),
            }
            for row in results
        ]
    except Exception as e:
        return []


@router.get("/top-losers")
async def get_top_losers(limit: int = Query(20, le=100)) -> List[Dict[str, Any]]:
    """Get top losers based on real-time price changes from Zerodha Kite WebSocket.

    Returns stocks with highest negative price changes today.
    """
    try:
        conn = get_db_connection()
        cur = conn.cursor()

        # Get real-time losers from WebSocket data
        # Calculate change_percent from close vs last_price
        cur.execute("""
            SELECT
                i.tradingsymbol as symbol,
                i.name,
                r.last_price::float as price,
                CASE
                    WHEN r.close > 0 THEN
                        ((r.last_price - r.close) / r.close * 100)::float
                    ELSE 0
                END as change,
                CASE
                    WHEN r.close > 0 THEN
                        LEAST(ABS((r.last_price - r.close) / r.close * 100) / 10.0, 1.0)
                    ELSE 0.5
                END as confidence
            FROM md.realtime_prices r
            JOIN md.instrument_tokens i ON i.instrument_token = r.instrument_token
            WHERE i.exchange = 'NSE'
              AND i.instrument_type = 'EQ'
              AND r.updated_at > NOW() - INTERVAL '10 minutes'
              AND r.close > 0
              AND r.last_price < r.close
            ORDER BY change ASC
            LIMIT %s
        """, (limit,))

        results = cur.fetchall()
        cur.close()
        conn.close()

        if not results:
            return []

        return [
            {
                "symbol": row[0],
                "name": row[1] or row[0],
                "change": round(float(row[3]) if row[3] else 0.0, 2),
                "confidence": round(float(row[4]) if row[4] else 0.0, 2),
                "price": round(float(row[2]) if row[2] else 0.0, 2),
            }
            for row in results
        ]
    except Exception as e:
        return []


@router.get("/{symbol}")
async def get_stock_data(symbol: str) -> Dict[str, Any]:
    """Get current stock data including price and change."""
    conn = get_db_connection()
    cur = conn.cursor()

    # Get latest price and previous day's price
    cur.execute("""
        WITH latest_prices AS (
            SELECT
                e.symbol,
                s.name,
                e.close,
                e.volume,
                s.market_cap,
                e.trade_date,
                LAG(e.close) OVER (ORDER BY e.trade_date) as prev_close
            FROM md.eod_prices e
            INNER JOIN md.symbols s ON e.symbol = s.symbol AND e.exchange = s.exchange
            WHERE e.symbol = %s
            AND e.exchange = 'NSE'
            ORDER BY e.trade_date DESC
            LIMIT 2
        )
        SELECT
            symbol,
            name,
            close as price,
            COALESCE(close - prev_close, 0) as change,
            CASE
                WHEN prev_close > 0 THEN ((close - prev_close) / prev_close * 100)
                ELSE 0
            END as change_percent,
            volume,
            market_cap
        FROM latest_prices
        ORDER BY trade_date DESC
        LIMIT 1
    """, (symbol,))

    result = cur.fetchone()
    cur.close()
    conn.close()

    if not result:
        raise HTTPException(status_code=404, detail=f"Stock {symbol} not found")

    return {
        "symbol": result[0],
        "name": result[1] or result[0],
        "price": round(float(result[2]) if result[2] else 0.0, 2),
        "change": round(float(result[3]) if result[3] else 0.0, 2),
        "changePercent": round(float(result[4]) if result[4] else 0.0, 2),
        "volume": int(result[5]) if result[5] else 0,
        "marketCap": int(result[6]) if result[6] else 0,
    }


@router.get("/{symbol}/history")
async def get_stock_history(
    symbol: str,
    days: int = Query(365, le=730, description="Number of days of history (max 730)"),
    start_date: Optional[str] = None,
    end_date: Optional[str] = None,
) -> List[Dict[str, Any]]:
    """Get historical stock data for the specified number of days.

    Returns OHLCV data for charting and analysis.
    If start_date/end_date are provided, they take precedence over days.
    """
    conn = get_db_connection()
    cur = conn.cursor()

    if start_date or end_date:
        filters = ["symbol = %s", "exchange = 'NSE'"]
        params: List[Any] = [symbol]

        if start_date:
            filters.append("trade_date >= %s")
            params.append(start_date)
        if end_date:
            filters.append("trade_date <= %s")
            params.append(end_date)

        where_sql = " AND ".join(filters)
        cur.execute(
            f"""
            SELECT
                trade_date,
                open,
                high,
                low,
                close,
                volume
            FROM md.eod_prices
            WHERE {where_sql}
            ORDER BY trade_date ASC
            """,
            tuple(params),
        )
    else:
        cur.execute("""
            SELECT
                trade_date,
                open,
                high,
                low,
                close,
                volume
            FROM md.eod_prices
            WHERE symbol = %s
            AND exchange = 'NSE'
            AND trade_date >= CURRENT_DATE - INTERVAL '%s days'
            ORDER BY trade_date ASC
        """, (symbol, days))

    results = cur.fetchall()
    cur.close()
    conn.close()

    if not results:
        raise HTTPException(
            status_code=404,
            detail=f"No historical data found for {symbol}"
        )

    return [
        {
            "date": row[0].isoformat(),
            "open": round(float(row[1]) if row[1] else 0.0, 2),
            "high": round(float(row[2]) if row[2] else 0.0, 2),
            "low": round(float(row[3]) if row[3] else 0.0, 2),
            "close": round(float(row[4]) if row[4] else 0.0, 2),
            "volume": int(row[5]) if row[5] else 0,
        }
        for row in results
    ]


@router.get("/realtime/all")
async def get_realtime_prices(limit: int = Query(50, le=200)) -> List[Dict[str, Any]]:
    """Get real-time prices for all tracked symbols from Zerodha Kite WebSocket."""
    try:
        conn = get_db_connection()
        cur = conn.cursor()

        cur.execute("""
            SELECT
                i.tradingsymbol as symbol,
                r.last_price::float as last_price,
                r.volume::bigint as volume,
                r.open::float as open,
                r.high::float as high,
                r.low::float as low,
                r.close::float as close,
                r.change_percent::float as change_percent,
                r.updated_at::text as updated_at
            FROM md.realtime_prices r
            JOIN md.instrument_tokens i ON i.instrument_token = r.instrument_token
            WHERE i.exchange = 'NSE'
              AND i.instrument_type = 'EQ'
              AND r.updated_at > NOW() - INTERVAL '5 minutes'
            ORDER BY r.updated_at DESC
            LIMIT %s
        """, (limit,))

        results = cur.fetchall()
        cur.close()
        conn.close()

        return [
            {
                "symbol": row[0],
                "last_price": float(row[1]) if row[1] else 0.0,
                "volume": int(row[2]) if row[2] else None,
                "open": float(row[3]) if row[3] else 0.0,
                "high": float(row[4]) if row[4] else 0.0,
                "low": float(row[5]) if row[5] else 0.0,
                "close": float(row[6]) if row[6] else 0.0,
                "change_percent": float(row[7]) if row[7] else None,
                "updated_at": row[8],
            }
            for row in results
        ]
    except Exception as e:
        # Return empty list on error
        return []


@router.get("/search")
async def search_stocks(
    q: str = Query(..., min_length=1, description="Search query")
) -> List[Dict[str, Any]]:
    """Search for stocks by symbol or name."""
    conn = get_db_connection()
    cur = conn.cursor()

    search_term = f"%{q}%"
    cur.execute("""
        SELECT
            symbol,
            name,
            exchange
        FROM md.symbols
        WHERE active = TRUE
        AND (
            symbol ILIKE %s
            OR name ILIKE %s
        )
        ORDER BY
            -- Exact matches first
            CASE WHEN symbol = UPPER(%s) THEN 0 ELSE 1 END,
            -- Then prefix matches
            CASE WHEN symbol LIKE UPPER(%s) || '%%' THEN 0 ELSE 1 END,
            symbol
        LIMIT 20
    """, (search_term, search_term, q.upper(), q.upper()))

    results = cur.fetchall()
    cur.close()
    conn.close()

    return [
        {
            "symbol": row[0],
            "name": row[1] or row[0],
            "exchange": row[2],
        }
        for row in results
    ]

File: internal/database/stocks.go
package database

import (
	"context"
	"fmt"
)

// TopMover represents a top gainer or loser stock
type TopMover struct {
	Symbol     string  `json:"symbol"`
	Name       string  `json:"name"`
	Change     float64 `json:"change"`
	Confidence float64 `json:"confidence"`
	Price      float64 `json:"price"`
}

// RealtimePrice represents a stock's current market price
type RealtimePrice struct {
	Symbol        string   `json:"symbol"`
	LastPrice     float64  `json:"last_price"`
	Volume        *int64   `json:"volume"`
	Open          float64  `json:"open"`
	High          float64  `json:"high"`
	Low           float64  `json:"low"`
	Close         float64  `json:"close"`
	ChangePercent *float64 `json:"change_percent"`
	UpdatedAt     string   `json:"updated_at"`
}

// StockData represents detailed stock information
type StockData struct {
	Symbol        string  `json:"symbol"`
	Name          string  `json:"name"`
	Price         float64 `json:"price"`
	Change        float64 `json:"change"`
	ChangePercent float64 `json:"changePercent"`
	Volume        int64   `json:"volume"`
	MarketCap     float64 `json:"marketCap"`
}

// StockSearchResult represents a search result
type StockSearchResult struct {
	Symbol   string `json:"symbol"`
	Name     string `json:"name"`
	Exchange string `json:"exchange"`
}

// GetTopGainers returns top gaining stocks by change percentage
func (db *DB) GetTopGainers(ctx context.Context, limit int) ([]TopMover, error) {
	query := `
		SELECT
			rp.symbol,
			COALESCE(sc.name, rp.symbol) as name,
			COALESCE(rp.change_percent, 0) as change,
			0.7 as confidence,
			COALESCE(rp.last_price, 0) as price
		FROM md.realtime_prices rp
		LEFT JOIN md.stock_config sc ON sc.symbol = rp.symbol AND sc.exchange = COALESCE(rp.exchange, 'NSE')
		WHERE rp.change_percent IS NOT NULL AND rp.change_percent > 0
			AND rp.updated_at > NOW() - INTERVAL '1 day'
			AND rp.symbol IS NOT NULL
		ORDER BY rp.change_percent DESC
		LIMIT $1
	`
	rows, err := db.conn.QueryContext(ctx, query, limit)
	if err != nil {
		return nil, fmt.Errorf("failed to query top gainers: %w", err)
	}
	defer rows.Close()

	var results []TopMover
	for rows.Next() {
		var m TopMover
		if err := rows.Scan(&m.Symbol, &m.Name, &m.Change, &m.Confidence, &m.Price); err != nil {
			return nil, fmt.Errorf("failed to scan top gainer: %w", err)
		}
		results = append(results, m)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("rows iteration error: %w", err)
	}
	return results, nil
}

// GetTopLosers returns top losing stocks by change percentage
func (db *DB) GetTopLosers(ctx context.Context, limit int) ([]TopMover, error) {
	query := `
		SELECT
			rp.symbol,
			COALESCE(sc.name, rp.symbol) as name,
			COALESCE(rp.change_percent, 0) as change,
			0.7 as confidence,
			COALESCE(rp.last_price, 0) as price
		FROM md.realtime_prices rp
		LEFT JOIN md.stock_config sc ON sc.symbol = rp.symbol AND sc.exchange = COALESCE(rp.exchange, 'NSE')
		WHERE rp.change_percent IS NOT NULL AND rp.change_percent < 0
			AND rp.updated_at > NOW() - INTERVAL '1 day'
			AND rp.symbol IS NOT NULL
		ORDER BY rp.change_percent ASC
		LIMIT $1
	`
	rows, err := db.conn.QueryContext(ctx, query, limit)
	if err != nil {
		return nil, fmt.Errorf("failed to query top losers: %w", err)
	}
	defer rows.Close()

	var results []TopMover
	for rows.Next() {
		var m TopMover
		if err := rows.Scan(&m.Symbol, &m.Name, &m.Change, &m.Confidence, &m.Price); err != nil {
			return nil, fmt.Errorf("failed to scan top loser: %w", err)
		}
		results = append(results, m)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("rows iteration error: %w", err)
	}
	return results, nil
}

// GetRealtimePrices returns latest prices for multiple stocks
func (db *DB) GetRealtimePrices(ctx context.Context, limit int) ([]RealtimePrice, error) {
	query := `
		SELECT
			symbol,
			COALESCE(last_price, 0),
			volume,
			COALESCE(open, 0),
			COALESCE(high, 0),
			COALESCE(low, 0),
			COALESCE(close, 0),
			change_percent,
			COALESCE(updated_at::text, '')
		FROM md.realtime_prices
		WHERE symbol IS NOT NULL
			AND updated_at > NOW() - INTERVAL '1 day'
		ORDER BY updated_at DESC
		LIMIT $1
	`
	rows, err := db.conn.QueryContext(ctx, query, limit)
	if err != nil {
		return nil, fmt.Errorf("failed to query realtime prices: %w", err)
	}
	defer rows.Close()

	var results []RealtimePrice
	for rows.Next() {
		var p RealtimePrice
		if err := rows.Scan(&p.Symbol, &p.LastPrice, &p.Volume, &p.Open, &p.High, &p.Low, &p.Close, &p.ChangePercent, &p.UpdatedAt); err != nil {
			return nil, fmt.Errorf("failed to scan realtime price: %w", err)
		}
		results = append(results, p)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("rows iteration error: %w", err)
	}
	return results, nil
}

// GetRealtimePrice returns the latest price for a single stock
func (db *DB) GetRealtimePrice(ctx context.Context, symbol string) (*RealtimePrice, error) {
	query := `
		SELECT
			symbol,
			COALESCE(last_price, 0),
			volume,
			COALESCE(open, 0),
			COALESCE(high, 0),
			COALESCE(low, 0),
			COALESCE(close, 0),
			change_percent,
			COALESCE(updated_at::text, '')
		FROM md.realtime_prices
		WHERE symbol = $1
		LIMIT 1
	`
	var p RealtimePrice
	err := db.conn.QueryRowContext(ctx, query, symbol).Scan(
		&p.Symbol, &p.LastPrice, &p.Volume, &p.Open, &p.High, &p.Low, &p.Close, &p.ChangePercent, &p.UpdatedAt,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to get realtime price for %s: %w", symbol, err)
	}
	return &p, nil
}

// GetStockData returns detailed stock data
func (db *DB) GetStockData(ctx context.Context, symbol string) (*StockData, error) {
	query := `
		SELECT
			sc.symbol,
			COALESCE(sc.name, sc.symbol) as name,
			COALESCE(rp.last_price, 0) as price,
			COALESCE(rp.last_price - rp.close, 0) as change,
			COALESCE(rp.change_percent, 0) as change_percent,
			COALESCE(rp.volume, 0) as volume,
			0 as market_cap
		FROM md.stock_config sc
		LEFT JOIN md.realtime_prices rp ON rp.symbol = sc.symbol
		WHERE sc.symbol = $1
		LIMIT 1
	`
	var s StockData
	err := db.conn.QueryRowContext(ctx, query, symbol).Scan(
		&s.Symbol, &s.Name, &s.Price, &s.Change, &s.ChangePercent, &s.Volume, &s.MarketCap,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to get stock data for %s: %w", symbol, err)
	}
	return &s, nil
}

// SearchStocks searches for stocks by symbol or name
func (db *DB) SearchStocks(ctx context.Context, query string) ([]StockSearchResult, error) {
	searchQuery := `
		SELECT symbol, COALESCE(name, symbol), exchange
		FROM md.stock_config
		WHERE active = true
			AND (symbol ILIKE $1 OR name ILIKE $1)
		ORDER BY
			CASE WHEN symbol ILIKE $2 THEN 0 ELSE 1 END,
			symbol
		LIMIT 20
	`
	pattern := "%" + query + "%"
	exactPattern := query + "%"
	rows, err := db.conn.QueryContext(ctx, searchQuery, pattern, exactPattern)
	if err != nil {
		return nil, fmt.Errorf("failed to search stocks: %w", err)
	}
	defer rows.Close()

	var results []StockSearchResult
	for rows.Next() {
		var r StockSearchResult
		if err := rows.Scan(&r.Symbol, &r.Name, &r.Exchange); err != nil {
			return nil, fmt.Errorf("failed to scan search result: %w", err)
		}
		results = append(results, r)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("rows iteration error: %w", err)
	}
	return results, nil
}
